### ISSUE TITLE
Distance sorting not working properly

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12755]
### Describe your problem!

A user contacted us that after upgrading to 2022.02.13 he has problems sorting caches by distance.

I am a bit clueless as the screenshots really seem to show a bug:
![Screenshot_2022-02-15-20-16-41-570_cgeo geocaching](https://user-images.githubusercontent.com/949669/154153715-3b47df9f-b77e-45ab-8b40-c998df15ef3b.jpg)
![Screenshot_2022-02-14-23-50-55-877_cgeo geocaching](https://user-images.githubusercontent.com/949669/154153744-595431b3-ca04-454c-b2f1-7c06db9659c3.jpg)

Any idea how to debug or get more info?
Or even an idea what might go on here?

### How to reproduce?

Unclear

### Reproducible

Unclear

### c:geo Version

2022.02.13

### System information

```text
c:geo version: 2022.02.13

Device:
-------
- Device type: Redmi Note 9S (curtana_eea, Redmi)
- Available processors: 8
- Android version: 11
- Android build: RKQ1.200826.002 test-keys
- Screen resolution: 1080x2168px (392x788dp)
- Pixel density: 2.75
- System font scale: 1.0 / used scale: 1.0
- Sailfish OS detected: false
- Google Play services: enabled - 21.48.15 (150400-414534850)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: absent
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:
-------
- Settings: v9, Count:157
- Set language: nl_NL (system default)
- System date format: dd-MM-y
- Time zone: GMT+01:00
- Debug mode active: no
- Last backup: 31 jan. 07:26
- Routing mode: Off
- Live map mode: true
- OSM multi-threading: false / threads: 1
- Map: Google: Map
  - Id: cgeo.geocaching.maps.google.v2.GoogleMapProvider$GoogleMapSource
  - Atts: none
  - Theme: none

Filters:
-------
- Hide waypoints: -
- LIVE: Geen ([:inconclusive=false:advanced=true])
- OFFLINE: Geen ([:inconclusive=false:advanced=true])

Stored Filters (#5):
- Gearchiveerd: [Gearchiveerd:inconclusive=false:advanced=true]AND(status:found_no:exclude_active:exclude_disabled))
- Meer dan 85 km: [Meer dan 85 km:inconclusive=false:advanced=true]AND(distance:87.0:-:coord=-:use_current_pos=true))
- NFont NSaved: [NFont NSaved:inconclusive=false:advanced=true]AND(status:found_no:exclude_disabled:exclude_archived;type:TRADITIONAL))
- Nfound NSaved Tradies: [Nfound NSaved Tradies:inconclusive=false:advanced=true]AND(status:found_no:stored_no:exclude_disabled:exclude_archived;type:TRADITIONAL))
- NietOpgeslagen Tradies: [NietOpgeslagen Tradies:inconclusive=false:advanced=true]AND(type:TRADITIONAL;status:found_no:stored_no:exclude_disabled:exclude_archived;distance:-:204.0:coord=-:use_current_pos=true;difficulty_terrain:d=1.0:d=4.0:t=1.0:t=4.0))

Services:
-------
- Geocaching sites enabled:
   geocaching.com: Logged in (Ingelogd) / PREMIUM
   Geocaching.com Adventure Lab
- Geocaching.com date format: dd-MM-yyyy
- Routing: external / BRouter installed: false
- Installed c:geo plugins:  none

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (73,1 GB free) v3 internal isDir(7 entries)
- Legacy User storage c:geo dir: /data/user/0/cgeo.geocaching (73,1 GB free) v3 internal isDir(7 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (73,1 GB free) v3 external non-removable isDir(4257 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (73,1 GB free) v3 internal isDir(0 entries)
- Public Folders: #11
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=25, dirs:>=10, totalFileSize:>=3,9 GB, free space: 73,1 GB, files on device: 13511677)
  - OFFLINE_MAPS: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=25, dirs:>=10, totalFileSize:>=3,9 GB, free space: 73,1 GB, files on device: 13511677)
  - OFFLINE_MAP_THEMES: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=25, dirs:>=10, totalFileSize:>=3,9 GB, free space: 73,1 GB, files on device: 13511677)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:1, dirs:0, totalFileSize:24,9 KB, free space: 73,1 GB, files on device: 13511677)
  - GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:15, dirs:0, totalFileSize:1004,8 MB, free space: 73,1 GB, files on device: 13511677)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:4, dirs:2, totalFileSize:261,1 MB, free space: 73,1 GB, files on device: 13511677)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:1, dirs:0, totalFileSize:0 B, free space: 73,1 GB, files on device: 13511677)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 73,1 GB, files on device: 13511677)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:7, dirs:1, totalFileSize:81,2 KB, free space: 73,1 GB, files on device: 13511677)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 73,1 GB, files on device: 13511677)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 73,1 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- UNUSED_TRACK: null
- Persisted Uri Permissions: #1
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (19 apr. 2021 17:24):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v98, Size:124,1 MB) on system internal storage

--- End of system information ---
```

### Additional Information

Support ticket 259893

**Comment by Lineflyer** [comment]
Could it be, that coincidentally owned caches are not sorted properly?

**Comment by Lineflyer** [comment]
Forget it...others also involved but no so drastically.

**Comment by eddiemuc** [comment]
Is this a single case, and is it reproducible?

The only thing coming to my mind right now causing this is if somehow c:geo does not get the (correct) current position of the user for the sorting (but for the distance-per-cache-calculation it does).

**Comment by Lineflyer** [comment]
I tried my best to reproduce but failed. So far it seems a single case.

**Comment by Lineflyer** [comment]
I asked the user a few questions to potentially find a clue on this:
- Is this faulty sorting stable or does it change while viewing the list or e.g. when scrolling down and up again?
- Do you also have this problem if you reverse the sorting (by selecting "distance" again)?
- Do you have the same problem if you set the sorting to "Automatic" (which should also be distance sorting)?
- Does it also happen for other sortings (e.g when sorting by favorite points)?

**Comment by Lineflyer** [comment]
Send a reminder to that user as he did not reply to the questions yet.
If no reply comes in suggest to close this.

**Comment by Lineflyer** [comment]
Triggered by another support mail I finally found the possible reason or at least a very similar case:

- If you initially open a list sorted by distance its properly sorted
- If you however start moving the distances in the list will be updated accordingly but the sorting does not change, meaning it always stays sorted by the initial distance when opening the list not taking into account the changed distance

**Comment by eddiemuc** [comment]
If I understand the feature request correctly, the list should automatically be resorted periodically if chosen sort is "distance to own position". Correct?

Follow-up-questions:
* How often shall this happen? E.g. every 10 seconds? Every Minute? Configurable?
* What about results of online searches (e.g. search by keyword)? Should online search be retriggered automatically (which may result in different result lists of course - if you have moved inbetween then caches not part of previous result may become part of result now)

**Comment by SammysHP** [comment]
In my opinion:

> the list should automatically be resorted periodically if chosen sort is "distance to own position".

No, only when the sort button is tapped (again), which has to be done twice, of course. Otherwise the items would jump around all the time. Also, don't we calculate the distance around the search center?

> What about results of online searches (e.g. search by keyword)? Should online search be retriggered automatically (which may result in different result lists of course - if you have moved inbetween then caches not part of previous result may become part of result now)

No new search should be triggered. We can think about adding "pull down to refresh" to lists, but that's worth another issue.

**Comment by Lineflyer** [comment]
> If I understand the feature request correctly,[...]

This sounds like it never worked, but actually I remember driving around some years ago with a list opened and the caches changed their list position while doing so. But I have to admit its long time ago and I don't remember 100% as I usually don't have this use case.

**Comment by eddiemuc** [comment]
> > If I understand the feature request correctly,[...]
> 
> This sounds like it never worked, ...

Admittetly, I don't know either if something like this worked before or not. But regardless of whether previous behaviour should be restored or a new request should be implemented -> it makes  sense imho to specify upfront what is wanted in the end.

**Comment by eddiemuc** [comment]
> In my opinion:
> 
> > the list should automatically be resorted periodically if chosen sort is "distance to own position".
> 
> No, only when the sort button is tapped (again), which has to be done twice, of course. Otherwise the items would jump around all the time. Also, don't we calculate the distance around the search center?
> 
> > What about results of online searches (e.g. search by keyword)? Should online search be retriggered automatically (which may result in different result lists of course - if you have moved inbetween then caches not part of previous result may become part of result now)
> 
> No new search should be triggered. We can think about adding "pull down to refresh" to lists, but that's worth another issue.

This would mean to leave everything as is -> correct?

> Also, don't we calculate the distance around the search center?

You mean the distance displayed in each cache row? afaik this is usually calculated from the current position. I don't know if this changes if the list it the result from a search around a search center other than current pos (e.g. result of an address search)

**Comment by fm-sys** [comment]
> 
> This sounds like it never worked, but actually I remember driving around some years ago with a list opened and the caches changed their list position while doing so.

Yes, I can confirm that it definitely was working some months ago (summer 2021). Everytime the distance to each cache was recalculated, it also had triggered a sort.

>What about results of online searches (e.g. search by keyword)?

The results should be resorted properly as well, but no network requests are expected by the user IMHO.

**Comment by Lineflyer** [comment]
Feedback by user:

I have seen activity on this report on github. To further clarify.  This function has worked for quite a long time. I have used it for at least the past two years.  The caches names do indeed jump up and down the viewable list as you get closer or further away from them.

 

I still have version 2021.12.24 on one of my devices, and tested it this morning. Took a ride and checked the list when I started, and while driving. The list sorted by distance when opened. As I was driving around the list resorted as the distances changed.

 

To replicate: open program

                        Stored

                        Show as list

                        Sort distance

 

I hope this helps, as I like this ability and use it often
